lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp

```diff
@@ -157,6 +158,244 @@ TypeSystemSwiftTypeRef::CanonicalizeSugar(swift::Demangle::Demangler &dem,
+    llvm::StringRef mangled_typename) {
+  if (mangled_typename.empty())
+    return {};
+
+  swift::Demangle::Demangler dem;
+  auto *type_node =
+      swift_demangle::GetDemangledTypeMangling(dem, mangled_typename);
+  if (!type_node)
+    return {};
+
+  auto flavor = SwiftLanguageRuntime::GetManglingFlavor(mangled_typename);
+  bool did_transform = false;
+  std::function<NodePointer(NodePointer)> transform_all_module_nodes =
+      [&](NodePointer node) -> NodePointer {
+    auto compiler_type = RemangleAsType(dem, node, flavor);
+    if (!compiler_type) {
+      LLDB_LOG(GetLog(LLDBLog::Types),
+               "[AdjustTypeForOriginallyDefinedInModule] Unexpected mangling "
+               "error when mangling adjusted node for type with mangled name "
+               "{0}",
+               mangled_typename);
+
+      return node;
+    }
+
+    // If this type has an @_originallyDefinedIn attribute, the compiler will
+    // emit a typealias whose name is prefixed by $ODI followed by the mangled
+    // name.
+    TargetSP target_sp(GetTargetWP().lock());
+    if (!target_sp)
+      return node;
+    ModuleList &module_list = target_sp->GetImages();
+    std::vector<ImportedDeclaration> decls;
+    module_list.FindImportedDeclarations(GetModule(),
+                                         compiler_type.GetMangledTypeName(),
+                                         decls, /*find_one=*/true);
+    if (decls.empty())
+      return node;
+
+    std::vector<lldb_private::CompilerContext> declContext =
+        decls[0].GetDeclContext();
+
+    lldbassert(!declContext.empty() &&
+               "Unexpected decl context for $ODI$ typealiased type");
+    if (declContext.empty())
+      return node;
+
+    auto module_context = declContext[0];
+
+    auto module_tranformer =
+        [&](NodePointer module_node,
+            CompilerContext module_context) -> NodePointer {
+      if (module_node->getText() == swift::MANGLING_MODULE_OBJC)
+        return module_node;
+
+      // If the mangled name's module and module context module match then
+      // there's nothing to do.
+      if (module_node->getText() == module_context.name)
+        return module_node;
+
+      // Otherwise this is a type who is originally defined in a separate
+      // module. Adjust the module name.
+      auto *adjusted_module_node = dem.createNodeWithAllocatedText(
+          Node::Kind::Module, module_context.name);
+      did_transform = true;
+      return adjusted_module_node;
+    };
+
+    NodePointer transformed_module =
+        TypeSystemSwiftTypeRef::TransformModuleName(
+            node, dem, [&](NodePointer module_node) {
+              return module_tranformer(module_node, module_context);
+            });
+
+    if (!transformed_module) {
+      return node;
+    }
+
+    NodePointer transformed_generic_types =
+        TypeSystemSwiftTypeRef::TransformBoundGenericTypes(
+            dem, transformed_module, transform_all_module_nodes);
```

Comment:
I'm not quite sold on the architecture of this algorithm yet. Why isn't this a single invocation of transform, that switches over the node type dispatches to local transformations?